### PR TITLE
Allow distinct clearing price per lease duration bucket

### PR DESCRIPTION
**Pull request by guggero**
This PR contains the client part of the change to switch over to distinct lease duration buckets for batches.

UPDATE: Made this non-breaking. Clients with batch version `0` won't be able to submit orders with a lease duration != 2016.

UPDATE2: Depends on #193 which was split off to make the server changes easier to split into multiple PRs.

### CODE REVIEW

**Comment by Roasbeef** [comment]
I think this is ready for review now?

**Comment by guggero** [comment]
Yes, has been for a while.

File: poolrpc/auctioneer.proto
Comment (halseth) [inline poolrpc/auctioneer.proto:990]: Call this `market open`?

File: poolrpc/auctioneer.proto
Comment (halseth) [inline poolrpc/auctioneer.proto:458]: I think we should avoid shifting the protos here. Feels like it could lead to some unexpected issues.

File: clientdb/batch_snapshot.go
Comment (halseth) [inline clientdb/batch_snapshot.go:441]: Great solution 👍

File: rpcserver.go
Comment (halseth) [inline rpcserver.go:1828]: I don't understand why this is needed. Won't the bid duration always match the duration for the batch it was matched in?

File: poolrpc/auctioneer.proto
Comment (halseth) [inline poolrpc/auctioneer.proto:1013]: wouldn't it be pretty easy to not break the API by adding this as a new field?

**Review by halseth — CHANGES_REQUESTED** [review]
First pass. Main question is whether we can avoid breaking the API.

File: poolrpc/auctioneer.proto
Comment (guggero) [reply poolrpc/auctioneer.proto:458]: Yeah, made all changes non-breaking for older clients.

File: rpcserver.go
Comment (guggero) [reply rpcserver.go:1828]: This is for legacy legacy orders where the durations were defined free style. When de-serializing those batch snapshots, the single clearing price is put into the map with the 2016 duration as the key. Otherwise we wouldn't know whether to look it up using the bid or ask duration (as they could differ before the uniform 2016 duration was introduced).

**Comment by guggero** [comment]
I was able to make everything non-breaking!
Still need to do some manual tests to confirm, but I think this PR is ready for a second look.

File: clientdb/batch_snapshot.go
Comment (wpaulino) [inline clientdb/batch_snapshot.go:514]: Could avoid the use of `EOF` markers by instead checking if the clearing price is non-zero signaling an old batch.

File: rpcserver.go
Comment (wpaulino) [inline rpcserver.go:248]: :+1:

**Review by wpaulino — COMMENTED** [review]
Non-breaking changes look good on first pass!

File: poolrpc/trader.proto
Comment (halseth) [inline poolrpc/trader.proto:878]: this breaks clients of the trader service.

File: poolrpc/auctioneer.proto
Comment (halseth) [inline poolrpc/auctioneer.proto:503]: Nice! 👍

File: order/rpc_parse.go
Comment (halseth) [inline order/rpc_parse.go:277]: should be removeed in later commit?

File: clientdb/batch_snapshot.go
Comment (halseth) [inline clientdb/batch_snapshot.go:390]: keep in mind that this might have only one element for new version as well, if only one bucket had matches.

**Review by halseth — CHANGES_REQUESTED** [review]
Great work on this PR! The latest version reads well to me. 

Also needs a rebase!

File: poolrpc/trader.proto
Comment (guggero) [reply poolrpc/trader.proto:898]: You're right, maybe someone is already using the RPC or REST interface. Deprecated the legacy format instead of replacing it.

File: clientdb/batch_snapshot.go
Comment (guggero) [reply clientdb/batch_snapshot.go:514]: Good idea! I thought it might screw with our unit tests if we always write a zero price. But it turns out this works well.

File: clientdb/batch_snapshot.go
Comment (guggero) [reply clientdb/batch_snapshot.go:390]: Removed, based on Wilmer's suggestion above.

File: clientdb/batch_snapshot_test.go
Comment (wpaulino) [inline clientdb/batch_snapshot_test.go:180]: Nit: Is `require.Equal` any better here?

**Review by wpaulino — APPROVED** [review]

**Review by halseth — APPROVED** [review]
Awesome, this turned out very clean in the end 👌 

Only nit from me is that I think we should make deprecated fields more clear by removing all but the deprecated message from their doc.

File: poolrpc/auctioneer.proto
Comment (guggero) [reply poolrpc/auctioneer.proto:1029]: Good idea, done :+1:

**Comment by guggero** [comment]
Rebased and addressed all comments.

Even though this is ready to be merged now, I think we should wait until the server part is close to being merged as well.

File: clientdb/batch_snapshot.go
Comment (halseth) [inline clientdb/batch_snapshot.go:527]: this is probably okay since these are just `uint32`, but triggers my spidey sense :P Could define these in the inner scope to make clear that each iteration is not overwriting the previous value.

**Review by halseth — APPROVED** [review]
Re-ACK ✅

File: clientdb/batch_snapshot.go
Comment (guggero) [reply clientdb/batch_snapshot.go:512]: Hmm, I thought I already replied to this, but apparently that comment got lost. I added the newline on purpose to make the return stand out a bit more since it's an early exit from the function.

File: clientdb/batch_snapshot_test.go
Comment (guggero) [reply clientdb/batch_snapshot_test.go]: Yes, the diff seems to be much cleaner :+1:

**Comment by guggero** [comment]
Replaced by #199.

**Pull request closed without merging**
